[Bug]: Flaky test for TestInternalIcebergCatalogService$TableTests.testTableTransaction:

### What happened?

```
2025-01-21T03:52:19.1457107Z [INFO] Tests run: 2, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.004 s - in org.apache.amoro.server.TestAmoroManagementConfValidator
2025-01-21T03:52:19.1474971Z 03:52:19.146 INFO org.apache.amoro.hive.HMSMockServer: -------------------------------------------------------------------------
2025-01-21T03:52:19.1475374Z 03:52:19.146 INFO org.apache.amoro.hive.HMSMockServer:     Shutdown HiveMetastoreServer 
2025-01-21T03:52:19.1475859Z 03:52:19.146 INFO org.apache.amoro.hive.HMSMockServer: -------------------------------------------------------------------------
2025-01-21T03:52:19.1476185Z 03:52:19.147 INFO org.apache.amoro.TestAms: Stop singleton mock AMS after testing.
2025-01-21T03:52:19.7173993Z [INFO] 
2025-01-21T03:52:19.7174361Z [INFO] Results:
2025-01-21T03:52:19.7174709Z [INFO] 
2025-01-21T03:52:19.7175003Z [ERROR] Errors: 
2025-01-21T03:52:19.7176611Z [ERROR]   TestInternalIcebergCatalogService$TableTests.testTableTransaction:174 » ServiceFailure Server error: PersistenceException: org.apache.ibatis.exceptions.PersistenceException: 
2025-01-21T03:52:19.7179631Z ### Error updating database.  Cause: org.apache.derby.shared.common.error.DerbySQLIntegrityConstraintViolationException: The statement was aborted because it would have caused a duplicate key value in a unique or primary key constraint or unique index identified by 'TABLE_RUNTIME_PK' defined on 'TABLE_RUNTIME'.
2025-01-21T03:52:19.7182079Z ### The error may exist in org/apache/amoro/server/persistence/mapper/TableMetaMapper.java (best guess)
2025-01-21T03:52:19.7183185Z ### The error may involve org.apache.amoro.server.persistence.mapper.TableMetaMapper.insertTableRuntime-Inline
2025-01-21T03:52:19.7184066Z ### The error occurred while setting parameters
2025-01-21T03:52:19.7186560Z ### SQL: INSERT INTO table_runtime (table_id, catalog_name, db_name, table_name, current_snapshot_id, current_change_snapshotId, last_optimized_snapshotId, last_optimized_change_snapshotId, last_major_optimizing_time, last_minor_optimizing_time, last_full_optimizing_time, optimizing_status_code, optimizing_status_start_time, optimizing_process_id, optimizer_group, table_config, pending_input, table_summary) VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)
2025-01-21T03:52:19.7190801Z ### Cause: org.apache.derby.shared.common.error.DerbySQLIntegrityConstraintViolationException: The statement was aborted because it would have caused a duplicate key value in a unique or primary key constraint or unique index identified by 'TABLE_RUNTIME_PK' defined on 'TABLE_RUNTIME'.
2025-01-21T03:52:19.7192807Z [INFO] 
2025-01-21T03:52:19.7193215Z [ERROR] Tests run: 605, Failures: 0, Errors: 1, Skipped: 8
```

please ref to the [ci](https://github.com/apache/amoro/actions/runs/12879647374/job/35907449916?pr=3409) for more detail

### Affects Versions

master

### What table formats are you seeing the problem on?

_No response_

### What engines are you seeing the problem on?

_No response_

### How to reproduce

_No response_

### Relevant log output

```shell

```

### Anything else

_No response_

### Are you willing to submit a PR?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [x] I agree to follow this project's Code of Conduct

--- a/amoro-ams/src/main/java/org/apache/amoro/server/table/DefaultTableService.java
+++ b/amoro-ams/src/main/java/org/apache/amoro/server/table/DefaultTableService.java
@@ -468,30 +468,30 @@ public class DefaultTableService extends PersistentBase implements TableService 
   }
 
   private void disposeTable(ServerTableIdentifier tableIdentifier) {
-    // Here, we first remove the tableRuntime before removing the tableIdentifier. This follows the
-    // reverse process of the syncTable() method, where we first add the tableIdentifier and then
-    // add the tableRuntime.
-    Optional.ofNullable(tableRuntimeMap.get(tableIdentifier.getId()))
-        .ifPresent(
-            tableRuntime -> {
-              try {
-                if (headHandler != null) {
-                  headHandler.fireTableRemoved(tableRuntime);
-                }
-                tableRuntime.dispose();
-                tableRuntimeMap.remove(
-                    tableIdentifier.getId()); // remove only after successful operation
-              } catch (Exception e) {
-                LOG.error("Error occurred while disposing table {}", tableIdentifier, e);
-              }
-            });
-    doAs(
-        TableMetaMapper.class,
-        mapper ->
-            mapper.deleteTableIdByName(
-                tableIdentifier.getCatalog(),
-                tableIdentifier.getDatabase(),
-                tableIdentifier.getTableName()));
+    doAsTransaction(
+        () ->
+            Optional.ofNullable(tableRuntimeMap.get(tableIdentifier.getId()))
+                .ifPresent(
+                    tableRuntime -> {
+                      try {
+                        if (headHandler != null) {
+                          headHandler.fireTableRemoved(tableRuntime);
+                        }
+                        tableRuntime.dispose();
+                        tableRuntimeMap.remove(
+                            tableIdentifier.getId()); // remove only after successful operation
+                      } catch (Exception e) {
+                        LOG.error("Error occurred while disposing table {}", tableIdentifier, e);
+                      }
+                    }),
+        () ->
+            doAs(
+                TableMetaMapper.class,
+                mapper ->
+                    mapper.deleteTableIdByName(
+                        tableIdentifier.getCatalog(),
+                        tableIdentifier.getDatabase(),
+                        tableIdentifier.getTableName())));
   }
 
   private static class TableIdentity {

--- a/amoro-ams/src/main/java/org/apache/amoro/server/table/DefaultTableService.java
+++ b/amoro-ams/src/main/java/org/apache/amoro/server/table/DefaultTableService.java
@@ -99,13 +99,22 @@ public class DefaultTableService extends PersistentBase implements TableService 
 
   @Override
   public void onTableDropped(InternalCatalog catalog, ServerTableIdentifier identifier) {
-    Optional.ofNullable(tableRuntimeMap.remove(identifier.getId()))
+    Optional.ofNullable(tableRuntimeMap.get(identifier.getId()))
         .ifPresent(
             tableRuntime -> {
-              if (headHandler != null) {
-                headHandler.fireTableRemoved(tableRuntime);
+              try {
+                if (headHandler != null) {
+                  headHandler.fireTableRemoved(tableRuntime);
+                }
+                tableRuntime.dispose();
+                tableRuntimeMap.remove(
+                    identifier.getId()); // remove only after successful operation
+              } catch (Exception e) {
+                LOG.error(
+                    "Error occurred while removing tableRuntime of table {}",
+                    identifier.getId(),
+                    e);
               }
-              tableRuntime.dispose();
             });
   }
 
@@ -255,7 +264,8 @@ public class DefaultTableService extends PersistentBase implements TableService 
     LOG.info("Syncing external catalogs took {} ms.", end - start);
   }
 
-  private void exploreExternalCatalog(ExternalCatalog externalCatalog) {
+  @VisibleForTesting
+  public void exploreExternalCatalog(ExternalCatalog externalCatalog) {
     final List<CompletableFuture<Set<TableIdentity>>> tableIdentifiersFutures =
         Lists.newArrayList();
     externalCatalog
@@ -458,6 +468,23 @@ public class DefaultTableService extends PersistentBase implements TableService 
   }
 
   private void disposeTable(ServerTableIdentifier tableIdentifier) {
+    // Here, we first remove the tableRuntime before removing the tableIdentifier. This follows the
+    // reverse process of the syncTable() method, where we first add the tableIdentifier and then
+    // add the tableRuntime.
+    Optional.ofNullable(tableRuntimeMap.get(tableIdentifier.getId()))
+        .ifPresent(
+            tableRuntime -> {
+              try {
+                if (headHandler != null) {
+                  headHandler.fireTableRemoved(tableRuntime);
+                }
+                tableRuntime.dispose();
+                tableRuntimeMap.remove(
+                    tableIdentifier.getId()); // remove only after successful operation
+              } catch (Exception e) {
+                LOG.error("Error occurred while disposing table {}", tableIdentifier, e);
+              }
+            });
     doAs(
         TableMetaMapper.class,
         mapper ->
@@ -465,14 +492,6 @@ public class DefaultTableService extends PersistentBase implements TableService 
                 tableIdentifier.getCatalog(),
                 tableIdentifier.getDatabase(),
                 tableIdentifier.getTableName()));
-    Optional.ofNullable(tableRuntimeMap.remove(tableIdentifier.getId()))
-        .ifPresent(
-            tableRuntime -> {
-              if (headHandler != null) {
-                headHandler.fireTableRemoved(tableRuntime);
-              }
-              tableRuntime.dispose();
-            });
   }
 
   private static class TableIdentity {

--- a/amoro-ams/src/main/java/org/apache/amoro/server/table/DefaultTableService.java
+++ b/amoro-ams/src/main/java/org/apache/amoro/server/table/DefaultTableService.java
@@ -468,30 +468,36 @@ public class DefaultTableService extends PersistentBase implements TableService 
   }
 
   private void disposeTable(ServerTableIdentifier tableIdentifier) {
-    doAsTransaction(
-        () ->
-            Optional.ofNullable(tableRuntimeMap.get(tableIdentifier.getId()))
-                .ifPresent(
-                    tableRuntime -> {
-                      try {
-                        if (headHandler != null) {
-                          headHandler.fireTableRemoved(tableRuntime);
+    TableRuntime existedTableRuntime = tableRuntimeMap.get(tableIdentifier.getId());
+    try {
+      doAsTransaction(
+          () ->
+              Optional.ofNullable(existedTableRuntime)
+                  .ifPresent(
+                      tableRuntime -> {
+                        try {
+                          if (headHandler != null) {
+                            headHandler.fireTableRemoved(tableRuntime);
+                          }
+                          tableRuntime.dispose();
+                          tableRuntimeMap.remove(
+                              tableIdentifier.getId()); // remove only after successful operation
+                        } catch (Exception e) {
+                          LOG.error("Error occurred while disposing table {}", tableIdentifier, e);
                         }
-                        tableRuntime.dispose();
-                        tableRuntimeMap.remove(
-                            tableIdentifier.getId()); // remove only after successful operation
-                      } catch (Exception e) {
-                        LOG.error("Error occurred while disposing table {}", tableIdentifier, e);
-                      }
-                    }),
-        () ->
-            doAs(
-                TableMetaMapper.class,
-                mapper ->
-                    mapper.deleteTableIdByName(
-                        tableIdentifier.getCatalog(),
-                        tableIdentifier.getDatabase(),
-                        tableIdentifier.getTableName())));
+                      }),
+          () ->
+              doAs(
+                  TableMetaMapper.class,
+                  mapper ->
+                      mapper.deleteTableIdByName(
+                          tableIdentifier.getCatalog(),
+                          tableIdentifier.getDatabase(),
+                          tableIdentifier.getTableName())));
+    } catch (Throwable t) {
+      tableRuntimeMap.putIfAbsent(tableIdentifier.getId(), existedTableRuntime);
+      throw t;
+    }
   }
 
   private static class TableIdentity {

--- a/amoro-ams/src/main/java/org/apache/amoro/server/table/TableOptimizingMetrics.java
+++ b/amoro-ams/src/main/java/org/apache/amoro/server/table/TableOptimizingMetrics.java
@@ -302,9 +302,11 @@ public class TableOptimizingMetrics {
   }
 
   public void unregister() {
-    registeredMetricKeys.forEach(globalRegistry::unregister);
-    registeredMetricKeys.clear();
-    globalRegistry = null;
+    if (globalRegistry != null) {
+      registeredMetricKeys.forEach(globalRegistry::unregister);
+      registeredMetricKeys.clear();
+      globalRegistry = null;
+    }
   }
 
   /**

--- a/amoro-ams/src/main/java/org/apache/amoro/server/table/TableOrphanFilesCleaningMetrics.java
+++ b/amoro-ams/src/main/java/org/apache/amoro/server/table/TableOrphanFilesCleaningMetrics.java
@@ -117,8 +117,10 @@ public class TableOrphanFilesCleaningMetrics {
   }
 
   public void unregister() {
-    registeredMetricKeys.forEach(globalRegistry::unregister);
-    registeredMetricKeys.clear();
-    globalRegistry = null;
+    if (globalRegistry != null) {
+      registeredMetricKeys.forEach(globalRegistry::unregister);
+      registeredMetricKeys.clear();
+      globalRegistry = null;
+    }
   }
 }

--- a/amoro-ams/src/main/java/org/apache/amoro/server/table/TableRuntime.java
+++ b/amoro-ams/src/main/java/org/apache/amoro/server/table/TableRuntime.java
@@ -174,6 +174,9 @@ public class TableRuntime extends StatedPersistentBase {
   }
 
   public void dispose() {
+    tableSummaryMetrics.unregister();
+    orphanFilesCleaningMetrics.unregister();
+    optimizingMetrics.unregister();
     tableLock.lock();
     try {
       doAsTransaction(
@@ -185,9 +188,6 @@ public class TableRuntime extends StatedPersistentBase {
     } finally {
       tableLock.unlock();
     }
-    optimizingMetrics.unregister();
-    orphanFilesCleaningMetrics.unregister();
-    tableSummaryMetrics.unregister();
   }
 
   public void beginPlanning() {

--- a/amoro-ams/src/main/java/org/apache/amoro/server/table/TableSummaryMetrics.java
+++ b/amoro-ams/src/main/java/org/apache/amoro/server/table/TableSummaryMetrics.java
@@ -233,9 +233,11 @@ public class TableSummaryMetrics {
   }
 
   public void unregister() {
-    registeredMetricKeys.forEach(globalRegistry::unregister);
-    registeredMetricKeys.clear();
-    globalRegistry = null;
+    if (globalRegistry != null) {
+      registeredMetricKeys.forEach(globalRegistry::unregister);
+      registeredMetricKeys.clear();
+      globalRegistry = null;
+    }
   }
 
   public void refresh(AbstractOptimizingEvaluator.PendingInput tableSummary) {

--- a/amoro-ams/src/test/java/org/apache/amoro/server/AMSManagerTestBase.java
+++ b/amoro-ams/src/test/java/org/apache/amoro/server/AMSManagerTestBase.java
@@ -59,6 +59,10 @@ public abstract class AMSManagerTestBase {
     EventsManager.dispose();
   }
 
+  protected DefaultCatalogManager catalogManager() {
+    return CATALOG_MANAGER;
+  }
+
   protected TableManager tableManager() {
     return TABLE_MANAGER;
   }

--- a/amoro-ams/src/test/java/org/apache/amoro/server/table/AMSTableTestBase.java
+++ b/amoro-ams/src/test/java/org/apache/amoro/server/table/AMSTableTestBase.java
@@ -249,6 +249,17 @@ public class AMSTableTestBase extends AMSServiceTestBase {
     tableService().exploreTableRuntimes();
   }
 
+  protected void dropTableOnly() {
+    if (externalCatalog == null) {
+      mixedTables.dropTableByMeta(tableMeta, true);
+      tableManager().dropTableMetadata(tableMeta.getTableIdentifier(), true);
+    } else {
+      String database = tableTestHelper.id().getDatabase();
+      String table = tableTestHelper.id().getTableName();
+      externalCatalog.dropTable(database, table, true);
+    }
+  }
+
   protected CatalogTestHelper catalogTestHelper() {
     return catalogTestHelper;
   }

--- a/amoro-ams/src/test/java/org/apache/amoro/server/table/TestSyncTableOfExternalCatalog.java
+++ b/amoro-ams/src/test/java/org/apache/amoro/server/table/TestSyncTableOfExternalCatalog.java
@@ -1,0 +1,274 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.amoro.server.table;
+
+import static org.apache.amoro.TableTestHelper.TEST_DB_NAME;
+import static org.apache.amoro.TableTestHelper.TEST_TABLE_NAME;
+import static org.apache.amoro.catalog.CatalogTestHelper.TEST_CATALOG_NAME;
+
+import org.apache.amoro.BasicTableTestHelper;
+import org.apache.amoro.ServerTableIdentifier;
+import org.apache.amoro.TableFormat;
+import org.apache.amoro.TableIDWithFormat;
+import org.apache.amoro.TableTestHelper;
+import org.apache.amoro.TestedCatalogs;
+import org.apache.amoro.api.TableIdentifier;
+import org.apache.amoro.catalog.CatalogTestHelper;
+import org.apache.amoro.hive.catalog.HiveCatalogTestHelper;
+import org.apache.amoro.hive.catalog.HiveTableTestHelper;
+import org.apache.amoro.server.catalog.ExternalCatalog;
+import org.apache.amoro.server.catalog.ServerCatalog;
+import org.apache.amoro.server.manager.MetricManager;
+import org.apache.amoro.server.metrics.MetricRegistry;
+import org.apache.amoro.server.persistence.PersistentBase;
+import org.apache.amoro.server.persistence.TableRuntimeMeta;
+import org.apache.amoro.server.persistence.mapper.TableMetaMapper;
+import org.junit.Assert;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.junit.runners.Parameterized;
+
+import java.util.List;
+
+@RunWith(Parameterized.class)
+public class TestSyncTableOfExternalCatalog extends AMSTableTestBase {
+
+  @Parameterized.Parameters(name = "{0}, {1}")
+  public static Object[] parameters() {
+    return new Object[][] {
+      {TestedCatalogs.hadoopCatalog(TableFormat.ICEBERG), new BasicTableTestHelper(true, true)},
+      {
+        new HiveCatalogTestHelper(TableFormat.ICEBERG, TEST_HMS.getHiveConf()),
+        new HiveTableTestHelper(true, true)
+      }
+    };
+  }
+
+  public TestSyncTableOfExternalCatalog(
+      CatalogTestHelper catalogTestHelper, TableTestHelper tableTestHelper) {
+    super(catalogTestHelper, tableTestHelper);
+  }
+
+  private final Persistency persistency = new Persistency();
+
+  private ExternalCatalog initExternalCatalog() {
+    ServerCatalog serverCatalog = catalogManager().getServerCatalog(TEST_CATALOG_NAME);
+    return catalogTestHelper().isInternalCatalog() ? null : (ExternalCatalog) serverCatalog;
+  }
+
+  /** Test synchronization after creating a table. */
+  @Test
+  public void testSynchronizationAfterCreateTable() {
+    // create table and sync table to Amoro server catalog
+    createTable();
+
+    // test list tables
+    List<TableRuntimeMeta> tableRuntimeMetaListAfterAddTable = persistency.getTableRuntimeMetas();
+    List<TableRuntimeMeta> tableRuntimeMetaListForOptimizerGroupAfterAddTable =
+        persistency.getTableRuntimeMetasForOptimizerGroup(defaultResourceGroup().getName());
+    Assert.assertEquals(
+        tableRuntimeMetaListForOptimizerGroupAfterAddTable.size(),
+        tableRuntimeMetaListAfterAddTable.size());
+
+    dropTable();
+    dropDatabase();
+  }
+
+  /** Test synchronization after dropping a table. */
+  @Test
+  public void testSynchronizationAfterDropTable() {
+    createTable();
+    // drop table and sync table to Amoro server catalog
+    dropTable();
+
+    // test list tables
+    List<TableRuntimeMeta> tableRuntimeMetaListAfterDropTable = persistency.getTableRuntimeMetas();
+    List<TableRuntimeMeta> tableRuntimeMetaListForOptimizerGroupAfterDropTable =
+        persistency.getTableRuntimeMetasForOptimizerGroup(defaultResourceGroup().getName());
+    Assert.assertEquals(
+        tableRuntimeMetaListForOptimizerGroupAfterDropTable.size(),
+        tableRuntimeMetaListAfterDropTable.size());
+
+    dropDatabase();
+  }
+
+  /**
+   * Test synchronization in the anomaly scenario where the tableIdentifier is remained in
+   * persistence but iceberg table does not exist.
+   */
+  @Test
+  public void testSynchronizationWithLegacyTableIdentifierAndNonExistingIcebergTable() {
+    // Simulate the anomaly scenario by only adding the table identifier in the persistent table
+    createDatabase();
+    ExternalCatalog externalCatalog = initExternalCatalog();
+
+    persistency.addTableIdentifier(
+        TEST_CATALOG_NAME, TEST_DB_NAME, TEST_TABLE_NAME, TableFormat.ICEBERG);
+
+    // Verify that the table not exists
+    List<TableIDWithFormat> tableIdentifierList =
+        catalogManager().getServerCatalog(TEST_CATALOG_NAME).listTables(TEST_DB_NAME);
+    Assert.assertEquals(0, tableIdentifierList.size());
+
+    // Verify that the tables do not match exactly in tableIdentifier and tableRuntime
+    List<TableRuntimeMeta> tableRuntimeMetaList = persistency.getTableRuntimeMetas();
+    List<ServerTableIdentifier> serverTableIdentifierList = tableManager().listManagedTables();
+    Assert.assertEquals(0, tableRuntimeMetaList.size());
+    Assert.assertEquals(1, serverTableIdentifierList.size());
+
+    // Verify that the synchronization works by firing exploreExternalCatalog
+    tableService().exploreExternalCatalog(externalCatalog);
+
+    List<TableRuntimeMeta> tableRuntimeMetaListAfterSync = persistency.getTableRuntimeMetas();
+    List<ServerTableIdentifier> serverTableIdentifierListAfterSync =
+        tableManager().listManagedTables();
+    Assert.assertEquals(0, tableRuntimeMetaListAfterSync.size());
+    Assert.assertEquals(0, serverTableIdentifierListAfterSync.size());
+
+    dropDatabase();
+  }
+
+  /**
+   * Test synchronization in the anomaly scenario where the tableRuntime is remained but iceberg
+   * table does not exist.
+   *
+   * <p>NOTE: exploreExternalCatalog does not work.
+   */
+  @Test
+  public void testSynchronizationWithLegacyTableRuntimeAndNonExistingIcebergTable() {
+    ExternalCatalog externalCatalog = initExternalCatalog();
+    // Simulate the anomaly scenario by only removing the table runtime in the persistent table
+    // after dropping table
+    createTable();
+    dropTableOnly();
+    persistency.deleteTableIdentifier(
+        serverTableIdentifier().getIdentifier().buildTableIdentifier());
+
+    // Verify that the table is dropped
+    List<TableIDWithFormat> tableIdentifierList =
+        catalogManager().getServerCatalog(TEST_CATALOG_NAME).listTables(TEST_DB_NAME);
+    Assert.assertEquals(0, tableIdentifierList.size());
+
+    // Verify that the tables do not match exactly in tableIdentifier and tableRuntime
+    List<ServerTableIdentifier> serverTableIdentifierList = tableManager().listManagedTables();
+    List<TableRuntimeMeta> tableRuntimeMetaListForOptimizerGroup =
+        persistency.getTableRuntimeMetasForOptimizerGroup(defaultResourceGroup().getName());
+    Assert.assertEquals(0, serverTableIdentifierList.size());
+    Assert.assertEquals(1, tableRuntimeMetaListForOptimizerGroup.size());
+
+    // Verify that the synchronization does not work by firing exploreExternalCatalog
+    tableService().exploreExternalCatalog(externalCatalog);
+
+    List<ServerTableIdentifier> serverTableIdentifierListAfterSync =
+        tableManager().listManagedTables();
+    List<TableRuntimeMeta> tableRuntimeMetaListForOptimizerGroupAfterSync =
+        persistency.getTableRuntimeMetasForOptimizerGroup(defaultResourceGroup().getName());
+    Assert.assertEquals(0, serverTableIdentifierListAfterSync.size());
+    Assert.assertEquals(1, tableRuntimeMetaListForOptimizerGroupAfterSync.size());
+
+    // The existed tableRuntime will prevent the table from being added again
+    Assert.assertThrows(IndexOutOfBoundsException.class, this::createTable);
+
+    MetricRegistry globalRegistry = MetricManager.getInstance().getGlobalRegistry();
+    globalRegistry.getMetrics().keySet().forEach(globalRegistry::unregister);
+    persistency.deleteTableRuntime(serverTableIdentifier().getId());
+    dropTable();
+    dropDatabase();
+  }
+
+  /**
+   * Test synchronization in the anomaly scenario where the tableIdentifier is remained in
+   * persistence and the tableRuntime is remained in memory but iceberg table does not exist.
+   */
+  @Test
+  public void
+      testSynchronizationWithLegacyTableIdentifierAndLegacyTableRuntimeAndNonExistingIcebergTable() {
+    ExternalCatalog externalCatalog = initExternalCatalog();
+    // Simulate the anomaly scenario by only removing the table runtime in the persistent table
+    // after dropping table
+    createTable();
+    dropTableOnly();
+    List<TableRuntimeMeta> tableRuntimeMetaListForOptimizerGroup =
+        persistency.getTableRuntimeMetasForOptimizerGroup(defaultResourceGroup().getName());
+    persistency.deleteTableRuntime(tableRuntimeMetaListForOptimizerGroup.get(0).getTableId());
+
+    // Verify that the table is dropped
+    List<TableIDWithFormat> tableIdentifierList =
+        catalogManager().getServerCatalog(TEST_CATALOG_NAME).listTables(TEST_DB_NAME);
+    Assert.assertEquals(0, tableIdentifierList.size());
+
+    // Verify that the tables do not match exactly in tableIdentifier and tableRuntime
+    List<TableRuntimeMeta> tableRuntimeMetaList = persistency.getTableRuntimeMetas();
+    List<ServerTableIdentifier> serverTableIdentifierList = tableManager().listManagedTables();
+    Assert.assertEquals(1, serverTableIdentifierList.size());
+    Assert.assertEquals(0, tableRuntimeMetaList.size());
+
+    // Verify that the synchronization works by firing exploreExternalCatalog
+    tableService().exploreExternalCatalog(externalCatalog);
+
+    List<TableRuntimeMeta> tableRuntimeMetaListAfterSync = persistency.getTableRuntimeMetas();
+    List<ServerTableIdentifier> serverTableIdentifierListAfterSync =
+        tableManager().listManagedTables();
+    Assert.assertEquals(0, serverTableIdentifierListAfterSync.size());
+    Assert.assertEquals(0, tableRuntimeMetaListAfterSync.size());
+
+    // Verify that recreating the table works
+    createTable();
+    tableRuntimeMetaListAfterSync = persistency.getTableRuntimeMetas();
+    serverTableIdentifierListAfterSync = tableManager().listManagedTables();
+    Assert.assertEquals(1, serverTableIdentifierListAfterSync.size());
+    Assert.assertEquals(1, tableRuntimeMetaListAfterSync.size());
+
+    dropTable();
+    dropDatabase();
+  }
+
+  private static class Persistency extends PersistentBase {
+    public void addTableIdentifier(
+        String catalog, String database, String tableName, TableFormat format) {
+      ServerTableIdentifier tableIdentifier =
+          ServerTableIdentifier.of(catalog, database, tableName, format);
+      doAs(TableMetaMapper.class, mapper -> mapper.insertTable(tableIdentifier));
+    }
+
+    public void deleteTableIdentifier(TableIdentifier tableIdentifier) {
+      doAs(
+          TableMetaMapper.class,
+          mapper ->
+              mapper.deleteTableIdByName(
+                  tableIdentifier.getCatalog(),
+                  tableIdentifier.getDatabase(),
+                  tableIdentifier.getTableName()));
+    }
+
+    public void deleteTableRuntime(Long tableId) {
+      doAs(TableMetaMapper.class, mapper -> mapper.deleteOptimizingRuntime(tableId));
+    }
+
+    public List<TableRuntimeMeta> getTableRuntimeMetasForOptimizerGroup(String optimizerGroup) {
+      return getAs(
+          TableMetaMapper.class,
+          mapper -> mapper.selectTableRuntimesForOptimizerGroup(optimizerGroup, null, null, null));
+    }
+
+    public List<TableRuntimeMeta> getTableRuntimeMetas() {
+      return getAs(TableMetaMapper.class, TableMetaMapper::selectTableRuntimeMetas);
+    }
+  }
+}

